### PR TITLE
fix: Copy modules and alias weights data when using tied weights

### PR DESCRIPTION
**Pull request by jeffkinnison**
The current tied weights implementation aliases torch modules between tied input features. Best practices seem to be deep copying the modules and then aliasing the underlying weights data (`Parameter.data`) between the copies.

Changes include:

- Tests updated to check for shared data pointers rather than shared module objects
- Refactored the tying code in `explain.utils` to be more accessible and work with input feature creation

### CODE REVIEW

**Comment by jeffkinnison** [comment]
The tests are all passing, but locally I found an `IndexError` that occurs in embedding layers when using tied weights. Looking into that and will update here with a fix.

**Comment by jeffkinnison** [comment]
After some more testing, it looks like the `IndexError` is likely an issue with `sample_ratio` rather than tied weights. Setting `sample_ratio: 0.1` in an `agnews` config seems to cause the issue regardless of whether to two text features are tied.

File: ludwig/explain/util.py [hunk 1]
```diff
@@ -83,29 +63,12 @@ def replace_layer_with_copy(feat: BaseFeature, target_layer: torch.nn.Module):
...
-    2. Replace the tensors in the copied encoder object with the tensors from the original encoder object, except for
-         the tensors in the target layer. We want to explain these tensors, so we want to keep them as deep copies.
-
     This approach ensures that at most 2 copies of the encoder object are in memory at any given time.
     """
     with torch.no_grad():
         # Get the original encoder object and a mapping from param names to the params themselves.
         orig_encoder_obj = feat.encoder_obj
-        orig_encoder_obj_state_dict = orig_encoder_obj.state_dict()
+        print(target_layer)
```
Comment (jppgks) [inline ludwig/explain/util.py:71]: leftover print statement from debugging?

File: ludwig/explain/util.py [hunk 2]
```diff
@@ -50,25 +49,6 @@ def get_feature_name(model: LudwigModel, target: str) -> str:
...
-    absolute_keys = []
-    for module_key, module_param in module.named_parameters():
-        for _, submodule_param in submodule.named_parameters():
-            if submodule_param.data_ptr() == module_param.data_ptr():
-                absolute_keys.append(module_key)
-                break
-    return absolute_keys
-
-
 def replace_layer_with_copy(feat: BaseFeature, target_layer: torch.nn.Module):
```
[the remark this answers is not in the document]
Comment (jeffkinnison) [reply ludwig/explain/util.py:52]: Good call! I'll check on that.

[the remark this answers is not in the document]
Comment (jeffkinnison) [reply ludwig/explain/util.py:52]: It seems like that works! I'll clean this up, then I think we should be good to go.

File: tests/integration_tests/test_input_feature_tied.py [hunk 3]
```diff
@@ -109,9 +109,15 @@ def test_tied_micro_level(input_feature_options):
 
     input_features = BaseModel.build_inputs(input_feature_configs=config_obj.input_features)
 
+    # Originally, we were aliasing encoder objects between two features and checking that the two features pointed to
+    # the same object. T
     if input_feature_options.tie_features:
         # should be same encoder
-        assert input_features["input_feature_1"].encoder_obj is input_features["input_feature_2"].encoder_obj
+        # assert input_features["input_feature_1"].encoder_obj is input_features["input_feature_2"].encoder_obj
```
Comment (justinxzhao) [inline tests/integration_tests/test_input_feature_tied.py:116]: nit: Delete?

**Review by justinxzhao — APPROVED** [review]

File: ludwig/explain/util.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jppgks) [reply ludwig/explain/util.py:52]: Is replace_layer_with_copy still used?

File: tests/ludwig/explain/test_util.py [hunk 4]
```diff
@@ -74,7 +75,7 @@ def test_replace_layer_with_copy(tmpdir):
     for param_name, param in input_feature_module.named_parameters():
         data_ptrs_before[param_name] = param.data_ptr()
 
-    keys_to_copy = get_absolute_module_key_from_submodule(input_feature_module, target_layer)
+    # keys_to_copy = get_absolute_module_key_from_submodule(input_feature_module, target_layer)
```
Comment (justinxzhao) [inline tests/ludwig/explain/test_util.py:78]: nit: Delete?

**Review by justinxzhao — APPROVED** [review]

**Pull request closed without merging**